README.md
## Function
Pada tugas kali ini anda akan mengembangkan beberapa fungsi dengan spesifikasi yang telah ditentukan.

### StudentDormitory
Butet baru saja bertemu dengan salah seorang dosen pemrograman yang memintanya untuk mengembangkan suatu program sederhana untuk mensimulasikan entitas-entitas di lingkungan kampus. Sebenarnya, sang dosen telah mengembangkan desain dasar dari simulator dan Butet hanya diminta untuk mengimplementasikannya. Sebagai langkah awal, simulator hanya akan menangani entitas mahasiswa (```student```), dan asrama (```dormitory```).

Berikut adalah rincian entitas dalam simulator:
1. Seorang ```student``` memiliki NIM (```id```), nama mahasiswa (```name```), angkatan (```year```),  program studi (```study program```), dan jenis kelamin (```gender```).
2. Sebuah ```dormitory``` memiliki nama asrama (```dormitory_name```), kapasitas atau muatan asrama (```capacity```), status (```status```), tipe asrama (```type```) dan jumlah penghuni (```residents```).
3. Terdapat dua status asrama, yaitu: ```Availabel``` apabila ```capacity > residents``` dan ```Not Available``` apabila ```capacity < residents```.
4. Setiap mahasiswa ditempatkan ke asrama berdasarkan yang berbeda berdasarkan jenis kelamin masing-masing mahasiswa. 

Simulator diharapkan memiliki fitur-fitur dasar sebagai berikut:
1. Menampilkan informasi dari seorang ```student``` ke ```stdout``` dan asrama ```dormitory``` yang akan di tempati masing-masing mahasiswa dengan menggunakan fungsi ```assign_students(...)```.
2. Menampilkan informasi penghuni (```mahasiswa```) yang tinggal di setiap asrama ```dormitory```. 
3. Menampilkan informasi  asrama ```dormitory``` dengan jumlah yang masih ```Available``` dan ```Not Available``` untuk dihuni mahasiswa.

Memperhatikan kompleksitas persoalan, Butet kemudian meminta bantuan dari sang sahabat, Ucok. Setelah mendengar penjelasan Butet, tanpa berpikir dua kali, ia pun bersedia membantu.

### Output Formats
Berikut adalah format keluaran untuk setiap upaya menampilkan informasi dari entitas:
1. ```assign_students(...)``` akan menampilkan untuk ```students``` id, nama mahasiswa, angkatan, program studi dan jenis kelamin dan untuk ```dormitory``` nama asrama, capacity, status dan type. Masing-masing data tersebut ditulis dalam satu baris keluaran dengan simbol hashtag ```#``` sebagai separator.
2. ```print_students_dorm(...)``` akan menampilkan nama asrama, kapasitas asrama, status dan tipe asrama. Keempat data tersebut ditulis dalam satu baris keluaran dengan simbol hashtag ```#``` sebagai separator.


### Compilation
Proses kompilasi harus menyertakan program dan ```StudentDormitory.c``` yang berisi implementasi dari spesifikasi. Berikut ini adalah contoh perintah untuk mengkompilasi ```t06_01.c```.
```gcc t06_01.c StudentDormitory.c -o t06_01
> 
```
Dari kompilasi akan dihasilkan sebuah berkas executable ```t06_01```.

### Case 1 (t06_01.c, 20 points)
Berikut adalah keluaran yang diharapkan.
```
Students:
12S20999#Wiro Sableng#2020#Information Systems#MALE
12S18019#Novita Hutapea#2018#Information Systems#FEMALE
12S19023#Irfan Siregar#2019#Information Systems#MALE
12S20030#Lamsihar Siahaan#2020#Information Systems#FEMALE
12S21056#Endang Siregar#2021#Information Systems#FEMALE
11319002#Asido Agripo Panjaitan#2019#Information Technology#MALE
11319014#Alex Sander Hutapea#2019#Information Technology#MALE
13319004#Sotar Dodo#2019#Computer Technology#MALE
21S20031#Aldi Sitorus#2020#Engineering Management#MALE
31S20020#Friskila Ariyanty#2020#Bioprocess Engineering#FEMALE

Dorms:
Pniel#10#Available#For Female
Silo#10#Available#For Male
Kapernaum#10#Available#For Male
Mamre#10#Available#For Female

```
### Case 2 (t06_02.c, 40 points)
Berikut adalah keluaran yang diharapkan.
```
Students:
12S20999#Wiro Sableng#2020#Information Systems#MALE
12S18019#Novita Hutapea#2018#Information Systems#FEMALE
12S19023#Irfan Siregar#2019#Information Systems#MALE
12S20030#Lamsihar Siahaan#2020#Information Systems#FEMALE
12S21056#Endang Siregar#2021#Information Systems#FEMALE
11319002#Asido Agripo Panjaitan#2019#Information Technology#MALE
11319014#Alex Sander Hutapea#2019#Information Technology#MALE
13319004#Sotar Dodo#2019#Computer Technology#MALE
21S20031#Aldi Sitorus#2020#Engineering Management#MALE
31S20020#Friskila Ariyanty#2020#Bioprocess Engineering#FEMALE

Dorms:
Pniel#10#Available#For Female
Silo#10#Available#For Male
Kapernaum#10#Available#For Male
Mamre#10#Available#For Female

Pniel#6#Available#For Female
12S18019#Novita Hutapea#2018#Information Systems
12S20030#Lamsihar Siahaan#2020#Information Systems
12S21056#Endang Siregar#2021#Information Systems
31S20020#Friskila Ariyanty#2020#Bioprocess Engineering

Silo#4#Available#For Male
12S20999#Wiro Sableng#2020#Information Systems
12S19023#Irfan Siregar#2019#Information Systems
11319002#Asido Agripo Panjaitan#2019#Information Technology
11319014#Alex Sander Hutapea#2019#Information Technology
13319004#Sotar Dodo#2019#Computer Technology
21S20031#Aldi Sitorus#2020#Engineering Management

Kapernaum#10#Available#For Male
0

Mamre#10#Available#For Female
0

```
### Case 3 (t06_03.c, 40 points)
Berikut adalah keluaran yang diharapkan.
```
Students:
12S20999#Wiro Sableng#2020#Information Systems#MALE
12S18019#Novita Hutapea#2018#Information Systems#FEMALE
12S19023#Irfan Siregar#2019#Information Systems#MALE
12S20030#Lamsihar Siahaan#2020#Information Systems#FEMALE
12S21056#Endang Siregar#2021#Information Systems#FEMALE
11319002#Asido Agripo Panjaitan#2019#Information Technology#MALE
11319014#Alex Sander Hutapea#2019#Information Technology#MALE
13319004#Sotar Dodo#2019#Computer Technology#MALE
21S20031#Aldi Sitorus#2020#Engineering Management#MALE
31S20020#Friskila Ariyanty#2020#Bioprocess Engineering#FEMALE

Dorms:
Pniel#3#Available#For Female
Silo#3#Available#For Male
Kapernaum#3#Available#For Male
Mamre#3#Available#For Female

Pniel#0#Not Available#For Female
12S18019#Novita Hutapea#2018#Information Systems
12S20030#Lamsihar Siahaan#2020#Information Systems
12S21056#Endang Siregar#2021#Information Systems

Silo#0#Not Available#For Male
12S20999#Wiro Sableng#2020#Information Systems
12S19023#Irfan Siregar#2019#Information Systems
11319002#Asido Agripo Panjaitan#2019#Information Technology

Kapernaum#0#Not Available#For Male
11319014#Alex Sander Hutapea#2019#Information Technology
13319004#Sotar Dodo#2019#Computer Technology
21S20031#Aldi Sitorus#2020#Engineering Management

Mamre#2#Available#For Female
31S20020#Friskila Ariyanty#2020#Bioprocess Engineering

```


## Reporting

Presentasikan pekerjaan anda dalam sebuah video. Pada presentasi:
1. Definisikan fungsi yang dikembangkan sebagai usulan solusi tugas ini. 
2. Berikan dasar pemikiran anda mengenai usulan solusi tersebut.
3. Jabarkan solusi anda.

**Note**: Semakin tajam argumen dan penjabaran anda semakin mudah penilaian dilakukan.

**Kriteria Video Presentasi**:
+ 1080p/30fps, min. 10 menit.
+ Suara jernih dan dapat dengan jelas terdengar.
+ Posting video anda di YouTube, di-set "Not For Kids" dengan visibility Unlisted.


## Submission
1. StudentDormitory.c
2. changelog.txt

## How to submit?
Please see https://youtu.be/g0BQ195-aWo


changelog.txt
mencoba
mencoba lagi
mencoba memperbaiki.

t06_02.c
// DO NOT EDIT
#include "StudentDormitory.h"
#include <stdio.h>
#include <stdlib.h>

int main(int _argc, char **_argv)
{
  struct student_t *std = malloc(12 * sizeof(struct student_t));
  struct dorm_t *dorm = malloc(4 * sizeof(struct dorm_t));
  struct student_dormitory_t *students_dorms = malloc(12 * sizeof(struct dorm_t));

  dorm[0] = create_dorm("Pniel", 10, A, FEMALE);
  dorm[1] = create_dorm("Silo", 10, A, MALE);
  dorm[2] = create_dorm("Kapernaum", 10, A, MALE);
  dorm[3] = create_dorm("Mamre", 10, A, FEMALE);

  
  std[0] =
      create_student("12S20999", "Wiro Sableng", "2020", "Information Systems", MALE);
  std[1] =
      create_student("12S18019", "Novita Hutapea", "2018", "Information Systems", FEMALE);
  std[2] =
      create_student("12S19023", "Irfan Siregar", "2019", "Information Systems", MALE);
  std[3] =
      create_student("12S20030", "Lamsihar Siahaan", "2020", "Information Systems", FEMALE);
  std[4] =
      create_student("12S21056", "Endang Siregar", "2021", "Information Systems", FEMALE);
  std[5] =
      create_student("11319002", "Asido Agripo Panjaitan", "2019", "Information Technology", MALE);
  std[6] =
      create_student("11319014", "Alex Sander Hutapea", "2019", "Information Technology", MALE);
  std[7] =
      create_student("13319004", "Sotar Dodo", "2019", "Computer Technology", MALE);
  std[8] =
      create_student("21S20031", "Aldi Sitorus", "2020", "Engineering Management", MALE);
  std[9] =
      create_student("31S20020", "Friskila Ariyanty", "2020", "Bioprocess Engineering", FEMALE);

  assign_students(students_dorms, dorm, std, 4, 10);

  print_students_dorm(students_dorms, 10, dorm, 4,std);

  return 0;
}

// DO NOT EDIT


StudentDormitory.c
// 12S21015 - Sitogab Antonio Octavianus Girsang
// 12S21049 - Jesika Audina Purba

#include "StudentDormitory.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct student_t create_student(char *_id, char *_name, char *_year, char *_study_program, enum gender_t _gender)
{
  struct student_t std;

  strcpy(std.id, _id);
  strcpy(std.name, _name);
  strcpy(std.year, _year);
  strcpy(std.study_program, _study_program);
  std.gender = _gender;

  return std;
}

struct dorm_t create_dorm(char *_dormitory_name, unsigned short _capacity, enum status_t _status, enum type_t _type)
{
  struct dorm_t dorm_;

  strcpy(dorm_.dormitory_name, _dormitory_name);
  dorm_.capacity = _capacity;
  dorm_.status = _status;
  dorm_.type = _type;

  return dorm_;
}

void print_students(struct student_t *_students, int size){

  printf("Students:\n");
  int i;
  for(i = 0; i < size; i++);
  {
    if (_students[i].gender == MALE)  
    {
      printf("%s#%s#%s#%s%s\n", _students[i].id, _students[i].name, _students[i].year, _students[i].study_program, _students[i].gender);
    } else if (_students[i].gender == FEMALE) {
      printf("%s#%s#%s#%s%s\n", _students[i].id, _students[i].name, _students[i].year, _students[i].study_program, _students[i].gender);
    }
  }

}

void print_dorms(struct dorm_t *_dorm, int size){
  printf("Dorms:\n");
  for (int i = 0; i < size; i++)
  {
    if (_dorm[i].status==A){
      if(_dorm[i].type==FOR_MALE){
        printf("%s#%d#Available#For Male\n", _dorm[i].dormitory_name, _dorm[i].capacity);
      } else  if(_dorm[i].type==FOR_FEMALE){
        printf("%s#%d#Available#For Female\n", _dorm[i].dormitory_name, _dorm[i].capacity);
      }
    } else if (_dorm[i].status==NA){
      if(_dorm[i].type==FOR_MALE){
        printf("%s#%d#Available#For Male\n", _dorm[i].dormitory_name, _dorm[i].capacity);
      } else if(_dorm[i].type==FOR_FEMALE){
        printf("%s#%d#Available#For Female\n", _dorm[i].dormitory_name, _dorm[i].capacity);
      }
    }
  }
}

void print_students_dorm(struct student_dormitory_t *students_dorms,int students_size, struct dorm_t *dorms, int droms_size, struct student_t *students)
{
   
    
    int cewe = 0;
    int cowo = 0;

    for (int dor = 0; dor < droms_size; dor++)
    {
        int lk = 0;
        int pr = 0;
        int batas = 0;

        for (int i = cowo; i < students_size; i++)
        {
            if (students[i].gender == 0 && dorms[dor].capacity != 0 && dorms[dor].type == 0)
            {
                dorms[dor].capacity--;
                batas++;
            }
        }

        for (int i = cewe; i < students_size; i++)
        {
            if (students[i].gender == 1 && dorms[dor].capacity != 0 && dorms[dor].type == 1)
            {
                dorms[dor].capacity--;
                batas++;
            }
        }

        printf("\n%s#%i#", dorms[dor].dormitory_name, dorms[dor].capacity);
        if (dorms[dor].capacity > 0)
        {
            printf("Available#");
        }
        if (dorms[dor].capacity == 0)
        {
            printf("Not Available#");
        }
        if (dorms[dor].type == 0)
        {
            printf("For Male\n");
            for (int dar = cowo; dar < students_size; dar++)
            {
                if (students[dar].gender == 0 && lk != batas)
                {
                    lk++;
                    cowo = dar + 1;
                    printf("%s#%s#%s#%s\n", students[dar].id, students[dar].name, students[dar].year, students[dar].study_program);
                }
            }
            if (lk == 0)
            {
                printf("0\n");
            }
        }
        else if (dorms[dor].type == 1)
        {
            printf("For Female\n");
            for (int dar = cewe; dar < students_size; dar++)
            {
                if (students[dar].gender == 1 && pr != batas)
                {
                    pr++;
                    cewe = dar + 1;
                    printf("%s#%s#%s#%s\n", students[dar].id, students[dar].name, students[dar].year, students[dar].study_program);
                }
            }
            if (pr == 0)
            {
                printf("0\n");
            }
        }
    }
}
void assign_students(struct student_dormitory_t *students_dorms, struct dorm_t *dorms, struct student_t *students, int size_dorms, int size_students)
{
  printf("Students:\n");
  for (int i = 0; i < size_students; i++)
  {
    printf("%s#%s#%s#%s#", students[i].id, students[i].name, students[i].year, students[i].study_program);
    if (students[i].gender == 0)
    {
      printf("MALE\n");
    }
    else if (students[i].gender == 1)
    {
      printf("FEMALE\n");
    }
  }
  printf("\nDorms:\n");
  for (int i = 0; i < size_dorms; i++)
  {
    printf("%s#%i#", dorms[i].dormitory_name, dorms[i].capacity);
    if (dorms[i].status == 0)
    {
      printf("Available#");
    }
    else if (dorms[i].status == 1)
    {
      printf("Not Available#");
    }
    if (dorms[i].type == 0)
    {
      printf("For Male\n");
    }
    else if (dorms[i].type == 1)
    {
      printf("For Female\n");
    }
  }
}

struct dorm_t check_dorm(struct dorm_t *dorms, int size, enum type_t gender)
{
  struct dorm_t temp;
  
  // CODE

  return temp;
}



t06_03.c
// DO NOT EDIT
#include "StudentDormitory.h"
#include <stdio.h>
#include <stdlib.h>

int main(int _argc, char **_argv)
{
    struct student_t *std = malloc(12 * sizeof(struct student_t));
    struct dorm_t *dorm = malloc(4 * sizeof(struct dorm_t));
    struct student_dormitory_t *students_dorms = malloc(12 * sizeof(struct dorm_t));

    dorm[0] = create_dorm("Pniel", 3, A, FEMALE);
    dorm[1] = create_dorm("Silo", 3, A, MALE);
    dorm[2] = create_dorm("Kapernaum", 3, A, MALE);
    dorm[3] = create_dorm("Mamre", 3, A, FEMALE);

    std[0] =
        create_student("12S20999", "Wiro Sableng", "2020", "Information Systems", MALE);
    std[1] =
        create_student("12S18019", "Novita Hutapea", "2018", "Information Systems", FEMALE);
    std[2] =
        create_student("12S19023", "Irfan Siregar", "2019", "Information Systems", MALE);
    std[3] =
        create_student("12S20030", "Lamsihar Siahaan", "2020", "Information Systems", FEMALE);
    std[4] =
        create_student("12S21056", "Endang Siregar", "2021", "Information Systems", FEMALE);
    std[5] =
        create_student("11319002", "Asido Agripo Panjaitan", "2019", "Information Technology", MALE);
    std[6] =
        create_student("11319014", "Alex Sander Hutapea", "2019", "Information Technology", MALE);
    std[7] =
        create_student("13319004", "Sotar Dodo", "2019", "Computer Technology", MALE);
    std[8] =
        create_student("21S20031", "Aldi Sitorus", "2020", "Engineering Management", MALE);
    std[9] =
        create_student("31S20020", "Friskila Ariyanty", "2020", "Bioprocess Engineering", FEMALE);

    assign_students(students_dorms, dorm, std, 4, 10);

    print_students_dorm(students_dorms, 10, dorm, 4,std);

    return 0;
}

// DO NOT EDIT
